File: solutions/java/code-carl/ch08_greedy/T5_LC0055_CanJump.java
/**
 * Practice template for:
 * - LeetCode 55. Jump Game
 *   https://leetcode.cn/problems/jump-game/
 */
public class T5_LC0055_CanJump {
    public boolean canJump(int[] nums) {
        int cover = 0;
        for (int i = 0; i <= cover; i++) {
            cover = Math.max(cover, i + nums[i]);
            if (cover >= nums.length - 1) {
                return true;
            }
        }
        return false;
    }
}

File: solutions/code-carl/ch08_greedy/t5_lc0055_can_jump.py
"""LeetCode 55. Jump Game
https://leetcode.cn/problems/jump-game/
"""
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if len(nums) <= 1:
            return True

        idx: int = 0
        cover: int = 0

        while idx <= cover:
            cover = max(cover, idx + nums[idx])
            if cover >= len(nums) - 1:
                return True
            idx += 1

        return False


if __name__ == "__main__":
    solution = Solution()
    assert solution.canJump([2, 3, 1, 1, 4]) == True
    assert solution.canJump([3, 2, 1, 0, 4]) == False
    print("All tests passed.")
